lib/flight_scheduler/message_sender.rb

```diff
@@ -0,0 +1,62 @@
+#==============================================================================
+# Copyright (C) 2020-present Alces Flight Ltd.
+#
+# This file is part of FlightSchedulerDaemon.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# FlightSchedulerDaemon is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with FlightSchedulerDaemon. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightSchedulerDaemon, please visit:
+# https://github.com/openflighthpc/flight-scheduler-daemon
+#==============================================================================
+
+module FlightScheduler
+  class MessageSender
+    def self.send(**opts)
+      new(**opts).tap(&:write)
+    end
+
+    attr_reader :opts, :task
+
+    def initialize(**opts)
+      @opts = opts
+    end
+
+    def wait
+      task.wait
+    end
+
+    def write
+      Async do |t|
+        @task = t
+
+        # Wait until a connection becomes available
+        connection = nil
+        until connection = FlightScheduler.app.connection
+          task.yield
```

Comment:
If there is a single task running `task.yield` will have `Async` immediately schedule that same task again.  This would result in very busy loop in which we repeatedly check if `FlightScheduler.app.connection` is truthy.

We can handle that case by using `task.sleep 0.1`.  Which will result in the task yielding and not being rescheduled for `0.1` seconds.

NOTE: I've given little thought to whether `0.1` is a sensible value here.

Reply:
https://github.com/openflighthpc/flight-scheduler-daemon/pull/16/commits/ad1ea13ab95aaadd34c4e9824fc507cf9170361f